### PR TITLE
Fix the centered SR onthefly code for R->C and nonhomogeneous parametes

### PR DESCRIPTION
**Pull request by inailuig**
The centered SR code has to precalculate〈Ok〉using a vjp.
However as pointed out in https://github.com/netket/netket/pull/652#discussion_r619709042 for R->C calculating the oks with a single vjp is not correct (the simplest way to see this is that these oks are real for all real parameters, but they should be complex).
The previous test example didn't actually test R->C, since the imaginary part of the result didn't depend on the real parameters.

This
- updates the test (need to double-check again that the example now actually covers everything)
- for centered=True (which needs the 〈Ok〉):
  - R->R and holomorphic C->C: no changes
  - R->C:  now uses separate vjps for the real and imag parts
  - inhomogeneous & non-holomorphic: converts all parameters to real and calls the R->C code
  - auto-detects R->R, R->C, assumes holomorphic C->C (with the option to specify that its not holomorphic) 
- centered=False: no changes
- Also partially adds support for non-holomorphic wavefunctions and tests them

### CODE REVIEW

**Comment by PhilipVinc** [comment]
Clemens, what bug does this fix? What is the problem?

**Comment by inailuig** [comment]
> Clemens, what bug does this fix? What is the problem?

@PhilipVinc  I updated the description

**Review by gcarleo — APPROVED** [review]

**Comment by PhilipVinc** [comment]
@inailuig let me know when you're done

File: netket/optimizer/sr/sr_onthefly_logic.py [hunk 1]
```diff
@@ -40,6 +40,14 @@ def O_vjp(forward_fn, params, samples, w):
     return jax.tree_map(sum_inplace, res)  # allreduce w/ MPI.SUM
 
 
+def O_vjp_rc(forward_fn, params, samples, w):
+    _, vjp_fun = jax.vjp(forward_fn, params, samples)
+    res_r, _ = vjp_fun(w)
+    res_i, _ = vjp_fun(-1.0j * w)
```
Comment (inailuig) [inline netket/optimizer/sr/sr_onthefly_logic.py:46]: I think it might be more efficient to relax the real parameters to complex and do just 1 backward pass, however its giving me errors in some cases. I suggest we merge this as is and maybe revisit it later on.

Comment (PhilipVinc) [reply netket/optimizer/sr/sr_onthefly_logic.py:46]: Interesting.
To be discussed

**Comment by PhilipVinc** [comment]
I have labelled this test file with a pytestmark so it is not tested with mpi.

However, when you have 5 minutes, could you actually label every testset with @onlyif_mpi and add a matmul check under mpi?
